probe/endpoint/ebpf.go

```diff
@@ -0,0 +1,229 @@
+	case "accept":
+		conn := ebpfConnection{
+			incoming:         true,
+			tuple:            tuple,
+			pid:              pid,
+			networkNamespace: networkNamespace,
+		}
+		t.openConnections[tuple.String()] = conn
+	case "close":
+		if deadConn, ok := t.openConnections[tuple.String()]; ok {
+			delete(t.openConnections, tuple.String())
+			t.closedConnections = append(t.closedConnections, deadConn)
+		} else {
+			log.Errorf("EbpfTracker error: unmatched close event: %s pid=%d netns=%s", tuple.String(), pid, networkNamespace)
+		}
+	}
+
+}
+
+func (t *EbpfTracker) run() {
+	stdout, err := t.cmd.StdoutPipe()
+	if err != nil {
+		log.Errorf("EbpfTracker error: %v", err)
+		return
+	}
+
+	if err := t.cmd.Start(); err != nil {
+		log.Errorf("EbpfTracker error: %v", err)
+		return
+	}
+
+	defer func() {
+		if err := t.cmd.Wait(); err != nil {
+			log.Errorf("EbpfTracker error: %v", err)
+		}
+
+		t.Lock()
+		t.dead = true
+		t.Unlock()
+	}()
+
+	reader := bufio.NewReader(stdout)
+	// skip first line of output table, containing the headers
+	if _, err := reader.ReadString('\n'); err != nil {
+		log.Errorf("EbpfTracker error: %v", err)
+		return
+	}
+
+	defer log.Infof("EbpfTracker exiting")
+
+	scn := bufio.NewScanner(reader)
+	for scn.Scan() {
+		txt := scn.Text()
+		line := strings.Fields(txt)
+
+		if len(line) != 7 {
+			log.Errorf("error parsing line %q", txt)
+			continue
+		}
+
+		eventType := line[0]
+
+		pid, err := strconv.Atoi(line[1])
+		if err != nil {
+			log.Errorf("error parsing pid %q: %v", line[1], err)
+			continue
+		}
+
+		sourceAddr := net.ParseIP(line[2])
+		if sourceAddr == nil {
+			log.Errorf("error parsing sourceAddr %q: %v", line[2], err)
+			continue
+		}
+
+		destAddr := net.ParseIP(line[3])
+		if destAddr == nil {
+			log.Errorf("error parsing destAddr %q: %v", line[3], err)
+			continue
+		}
+
+		sPort, err := strconv.ParseUint(line[4], 10, 16)
+		if err != nil {
+			log.Errorf("error parsing sourcePort %q: %v", line[4], err)
+			continue
+		}
+		sourcePort := uint16(sPort)
+
+		dPort, err := strconv.ParseUint(line[5], 10, 16)
+		if err != nil {
+			log.Errorf("error parsing destPort %q: %v", line[5], err)
+			continue
+		}
+		destPort := uint16(dPort)
+
+		networkNamespace := line[6]
+
+		tuple := fourTuple{sourceAddr.String(), destAddr.String(), sourcePort, destPort}
+
+		t.handleConnection(eventType, tuple, pid, networkNamespace)
+	}
+}
+
+// walkConnections calls f with all open connections and connections that have come and gone
+// since the last call to walkConnections
+func (t *EbpfTracker) walkConnections(f func(ebpfConnection)) {
+	t.Lock()
+	defer t.Unlock()
+
+	for _, connection := range t.openConnections {
+		f(connection)
+	}
+	for _, connection := range t.closedConnections {
+		f(connection)
+	}
+	t.closedConnections = t.closedConnections[:0]
+}
+
+func (t *EbpfTracker) hasDied() bool {
+	t.Lock()
+	defer t.Unlock()
+
+	return t.dead
+}
+
+func (t *EbpfTracker) initialize() {
+	t.initialized = true
```

Comment:
this does not seem necessary: t.initialized is only read & written from the reporter thread and not from the ebpf tracker "run()" thread.